enhancements/security/manual-cloud-credentials.md

```diff
@@ -0,0 +1,170 @@
+
+### Upgradable False Condition
+
+A field will be added to the cloud credential operator ConfigMap (possibly becoming a separate API type in another enhancement) storing a list of Updates (from ClusterVersion.AvailableUpdates) that have been reconciled, confirming that we have minted credentials and it is safe to upgrade to this release.
+
+```yaml
+  reconciledAvailableUpdates:
+  - version: 4.4.1
+    image: quay.io/openshift-release-dev/ocp-release:4.4.1-x86_64
+  - version: 4.4.2
+    image: quay.io/openshift-release-dev/ocp-release:4.4.2-x86_64
+```
+
+Entries will be added to this list by the CLI tooling below, however an administrator could do this by hand if desired.
+
+If the CCO is enabled per normal best practice install, this field will be ignored, and should be absent.
+
+However if CCO is disabled, it will compare the ClusterVersion.Spec.AvailableUpdates to this list of ReconciledAvailableUpdates. If *any* available update is not present in the reconciled list, the CCO will report Upgradable=False.
+
+### Initial Installation
+
+```bash
+$ oc adm credentials reconcile --manifests=/path/to/installer/manifests
+```
+
+This command would parse the CredentialsRequests in the manifests directory, mint credentials in the cloud using the local users environment variables or credentials file, store the status on the CredentialsRequsts (which are yet to be created), write Secret manifests containing the cloud credentials, and disable the CCO.
+
+### Upgrade Reconciliation
+
+```bash
+$ oc adm credentials reconcile --available-updates
+```
+
+This command will iterate all currently available updates in ClusterVersion.AvailableUpdates. For each given release image we will extract each CredentialsRequest and use the local users environment variables/credentials file to compare and adjust permissions as necessary.
+
+This command reconciles permissions *in an additive fashion*. We do not want to reconcile any reduction in permissions as that would affect the running components in the cluster right now. The command would print a warning if this situation is encountered, and it can be resolved after upgrade with the next command.
+
+For net new CredentialsRequests in the update, we have a problem. We cannot assume to store a secret in the target namespace as it may be for a component that does not exist in the currently running OpenShift release. To work around this problem we propose copying net new CredentialsRequests into the cluster (all CredentialsRequests live in the openshift-cloud-credential-operator namespace today), and introducing the concept of temporary secret storage. This would allow us to mint the credentials when we have the root creds available, grab the secrets which we typically only get at creation time and cannot retrieve later, and transfer them to the correct namespace when it is available.
+
+Each available update we process successfully is added to the reconciledAvailableUpdates list.
+
+### On-Going Reconciliation
```

Comment:
I think this isn't a good goal for this specific case. Users are hoping to controls the IAM using their tools, and providing something that won't be used a lot is imo over-engineering. We can try to solve the actual need here and add this cli based mirror of credential-operator later on.

Reply:
Per above I am unclear if the goal here is (1) I want to use my own tools to mint creds, (2) I don't want admin creds stored in my cluster, or (3) both. I will check in with PM.